bundles/automation/org.eclipse.smarthome.automation.provider.file/src/main/java/org/eclipse/smarthome/automation/internal/provider/file/AbstractFileProvider.java

```diff
@@ -0,0 +1,333 @@
+     */
+    protected Map<String, List<URL>> urls = new HashMap<String, List<URL>>();
+
+    protected List<ProviderChangeListener<E>> listeners = new ArrayList<ProviderChangeListener<E>>();
+
+    public void activate(Map<Parser<E>, Map<String, String>> parsers) {
+        for (Parser<E> parser : parsers.keySet()) {
+            addParser(parser, parsers.get(parser));
+        }
+        super.activate();
+        importResources(new File(WatchingDir));
+    }
+
+    @Override
+    public void deactivate() {
+        super.deactivate();
+        parsers.clear();
+        urls.clear();
+        providedObjectsHolder.clear();
+        listeners.clear();
+    }
+
+    @Override
+    protected String getSourcePath() {
+        return WatchingDir;
+    }
+
+    @Override
+    public void addProviderChangeListener(ProviderChangeListener<E> listener) {
+        synchronized (listeners) {
+            listeners.add(listener);
+        }
+
+    }
+
+    @Override
+    public Collection<E> getAll() {
+        synchronized (providedObjectsHolder) {
+            return providedObjectsHolder.values();
+        }
+    }
+
+    @Override
+    public void removeProviderChangeListener(ProviderChangeListener<E> listener) {
+        synchronized (listeners) {
+            listeners.remove(listener);
+        }
+    }
+
+    protected void addParser(Parser<E> parser, Map properties) {
+        String parserType = (String) properties.get(Parser.FORMAT);
+        parserType = parserType == null ? Parser.FORMAT_JSON : parserType;
+        parsers.put(parserType, parser);
+        List<URL> value = urls.get(parserType);
+        if (value != null && !value.isEmpty()) {
+            for (URL url : value) {
+                importFile(parserType, url);
+            }
+        }
+    }
+
+    protected void removeParser(Parser<E> parser, Map properties) {
+        String parserType = (String) properties.get(Parser.FORMAT);
+        parserType = parserType == null ? Parser.FORMAT_JSON : parserType;
+        parsers.remove(parserType);
+    }
+
+    protected void importResources(File file) {
+        if (file.exists()) {
+            File[] files = file.listFiles();
+            if (files != null) {
+                for (File f : files) {
+                    importResources(f);
+                }
+            } else {
+                try {
+                    URL url = file.toURI().toURL();
+                    String parserType = getParserType(url);
+                    importFile(parserType, url);
+                } catch (MalformedURLException e) {
+                    // can't happen for the 'file' protocol handler with a correctly formatted URI
+                }
+            }
+        }
+    }
+
+    protected void remove(File path) {
+        try {
+            URL url = path.toURI().toURL();
+            synchronized (providerPortfolio) {
+                removeElements(providerPortfolio.remove(url));
```

Comment:
Thanks for catching this - it is working only for files now, I assumed that when a directory is deleted, I'll receive events for all files in it. I'll fix it